### ISSUE TITLE
F# async issues

### ISSUE THREAD

**Issue by dsyme** [dotnet/interactive#711]
The  F# async programming model (and some others) expects SynchronizationContext.Current to be set for single-threaded UI-friendly programming models, to allow re-synchronization to the UI thread on async wait (using [SetSynchronizationContext](https://docs.microsoft.com/en-us/dotnet/api/system.threading.synchronizationcontext.setsynchronizationcontext?view=netcore-3.1))

The kernel should pre-set this on any execution threads to allow continuation requests to be pumped back through the kernel.  I think it can be implemented in terms of Task primitives.

This can be seen by the fact the following Async code never continues after the sleep:
```
async { 
     for i in 0 .. 20 do
           printfn "[%d] waiting..." System.Threading.Thread.CurrentThread.ManagedThreadId
           do! Async.Sleep 1000
           printfn "[%d] done waiting..." System.Threading.Thread.CurrentThread.ManagedThreadId
} |> Async.StartImmediate
```

(Aside - Starting the async on a background thread using `Async.Start` is possible but console output is then largely lost, and I also haven't had reliable results using `output.Update(content)` from a background thread - separately we should review the imlpementation of `Update` to check it is always marshalling the request back to a consistent thread or is otherwise able to be called from a background thread)

**Comment by dsyme** [comment]
@jonsequitur explained things to me in a call

1. There is no thread affinity in any of the UI actions.  

2. There is sequential execution of cells but these can start background async/tasks and wait for async/tasks 

3. There is a temporal affinity for console output and `display.Update` calls - these are only "noticed" by the UI if they happen before the end of cell execution - this applies especially in Jupyter Lab.  

So basically we need to document that console output and `display.Update` calls on background threads may be lost if they happen after cell execution completes.
